sdk/identity/azure-identity/src/chained_token_credential.cpp

```diff
@@ -66,10 +69,33 @@ ChainedTokenCredentialImpl::ChainedTokenCredentialImpl(
 AccessToken ChainedTokenCredentialImpl::GetToken(
     std::string const& credentialName,
     TokenRequestContext const& tokenRequestContext,
-    Context const& context) const
+    Context const& context)
 {
-  for (auto const& source : m_sources)
+  std::unique_lock<std::mutex> lock(m_cachingMutex, std::defer_lock);
+
+  constexpr size_t maxSentinel = std::numeric_limits<std::size_t>::max();
+
+  if (m_cacheSelectedCredential && m_SelectedCredentialIndex == maxSentinel)
   {
+    lock.lock();
+    // Check again in case another thread already set the index, and unlock the mutex.
+    if (m_SelectedCredentialIndex != maxSentinel)
+    {
+      lock.unlock();
+    }
+  }
+
+  size_t i = 0;
+  size_t end_index = m_sources.size();
```

Comment:
1. `end_index` => `endIndex`.
2. Just `end` name is more idiomatic with the iterators (`vec.begin()` and `vec.end()`). End Index - I am not sure if it is inclusive or not. But "Looping until `iter != end` is idiomatic).
```suggestion
  size_t end = m_sources.size();
```